### ISSUE TITLE
Subprocess timeout causes output to be returned as bytes in text mode

### ISSUE THREAD

**Issue by 273cfb33-81c1-4c10-a019-6866c4326b05** [python/cpython#87597]
BPO | [43431](https://bugs.python.org/issue43431)
--- | :---
Nosy | @giampaolo, @eryksun
Files | <li>[test_subprocess.py](https://bugs.python.org/file49856/test_subprocess.py "Uploaded as text/plain at 2021-03-08.05:41:45 by macdjord"): Demonstration of issue</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-03-08.05:41:46.323>
labels = ['3.8', 'type-bug', 'library', '3.9', '3.10']
title = 'Subprocess timeout causes output to be returned as bytes in text mode'
updated_at = <Date 2021-03-30.18:57:45.244>
user = 'https://bugs.python.org/macdjord'
```

bugs.python.org fields:
```python
activity = <Date 2021-03-30.18:57:45.244>
actor = 'eryksun'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2021-03-08.05:41:46.323>
creator = 'macdjord'
dependencies = []
files = ['49856']
hgrepos = []
issue_num = 43431
keywords = []
message_count = 3.0
messages = ['388257', '388265', '388272']
nosy_count = 3.0
nosy_names = ['giampaolo.rodola', 'eryksun', 'macdjord']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue43431'
versions = ['Python 3.8', 'Python 3.9', 'Python 3.10']
```

</p></details>

**Comment by macdjord** [comment]
Note: If you use `check=True` and the command fails, `.stdout` on the resulting `CalledProcessError` *is* a `str`, not `bytes`, making things even more confusing and inconsistent.

**Comment by LewisGaul** [comment]
There's another related bug here: when there's no output before the timeout is hit then you get the `stdout` and `stderr` fields containing `None` rather than empty str/bytes. I've adressed both issues in https://github.com/python/cpython/pull/95579.

**Comment by gpshead** [comment]
Unfortunately this counts as an API change as people will have written code since timeouts were introduced in 3.3 that blindly assumes TimeoutExpired.stdout/stderr are bytes.  So while it _might_ be a bug, it's also a breaking change and thus _feature_. Meaning it can't be backported. And likely needing a deprecation period.

**Comment by erlend-aasland** [comment]
The current behaviour was documented in ...:

- #97685

Unless there is a need for a breaking change, I suggest closing this as resolved.

**Comment by LewisGaul** [comment]
I don't think this should be closed, from my perspective this is just a bug, and should not have been documented as now changing it *would* be a breaking change (which would still be warranted in my opinion).

I raised a PR addressing this (https://github.com/python/cpython/pull/95579 - not sure why it was moved to draft state), and a discussion following the pushback (https://discuss.python.org/t/pr-review-request-decode-subprocess-output-in-text-mode-when-timeout-is-hit/19594/2) which didn't get any input.

This feels very much unfinished to me, and unfortunately remains an ugly wart for users who want to properly handle errors when using subprocess (especially in combination with mypy for type checking).

**Comment by gpshead** [comment]
I understand your frustration, but existing code already depends on the behavior given it has been this way for 11+ years now so it isn't "just a bug" in that sense - thus documenting the existing behavior to save everyone the trouble of discovering it the hard way.

I moved that PR to Draft as most PRs should really just start in Draft mode (a realtively new-to-github feature, we should use it more often) as that helps indicate that there are unresolved larger issues and it is unclear if the change as is, is even what we want.  See my https://github.com/python/cpython/pull/95579#issuecomment-1263819163 comment.

It sounds like there is a potential way forward based on the last comment from @zooba on the PR - keeping both the bytes and decoding to text on demand via a new TimeoutExpired attribute for the purpose.

I'm fine reopening this since you seem interested in working on it.
